redis/cluster.py

```diff
@@ -0,0 +1,1884 @@
+            except ResponseError as e:
+                message = e.__str__()
+                warnings.warn("ResponseError: {0}".format(message))
+                raise e
+            except BaseException as e:
+                warnings.warn("BaseException")
+                if connection:
+                    connection.disconnect()
+                raise e
+            finally:
+                if connection is not None:
+                    redis_node.connection_pool.release(connection)
+
+        raise ClusterError("TTL exhausted.")
+
+    def close(self):
+        try:
+            with self._lock:
+                if self.nodes_manager:
+                    self.nodes_manager.close()
+        except AttributeError:
+            # RedisCluster's __init__ can fail before nodes_manager is set
+            pass
+
+    def _process_result(self, command, res, **kwargs):
+        """
+        Process the result of the executed command.
+        The function would return a dict or a single value.
+
+        :type command: str
+        :type res: dict
+
+        `res` should be in the following format:
+            Dict<node_name, command_result>
+        """
+        if command in self.result_callbacks:
+            return self.result_callbacks[command](command, res, **kwargs)
+        elif len(res) == 1:
+            # When we execute the command on a single node, we can
+            # remove the dictionary and return a single response
+            return list(res.values())[0]
+        else:
+            return res
+
+
+class ClusterNode(object):
+    def __init__(self, host, port, server_type=None, redis_connection=None):
+        if host == 'localhost':
+            host = socket.gethostbyname(host)
+
+        self.host = host
+        self.port = port
+        self.name = get_node_name(host, port)
+        self.server_type = server_type
+        self.redis_connection = redis_connection
+
+    def __repr__(self):
+        return '[host={0},port={1},' \
+               'name={2},server_type={3},redis_connection={4}]' \
+            .format(self.host,
+                    self.port,
+                    self.name,
+                    self.server_type,
+                    self.redis_connection)
+
+    def __eq__(self, obj):
+        return isinstance(obj, ClusterNode) and obj.name == self.name
+
+
+class LoadBalancer:
+    """
+    Round-Robin Load Balancing
+    """
+
+    def __init__(self, start_index=0):
+        self.primary_to_idx = {}
+        self.start_index = start_index
+
+    def get_server_index(self, primary, list_size):
+        server_index = self.primary_to_idx.setdefault(primary,
+                                                      self.start_index)
+        # Update the index
+        self.primary_to_idx[primary] = (server_index + 1) % list_size
+        return server_index
+
+    def reset(self):
+        self.primary_to_idx.clear()
+
+
+class NodesManager:
+    def __init__(self, startup_nodes, from_url=False,
```

Comment:
Shouldn't from_url come first? If you think about it, it's likely that a user passes a string-based connection and then wants to populate the rest, as opposed to start_nodes and a from_url. 
If we take it a step further, perhaps both have no value, then it's checked and we raise an ValueError if things aren't set accordingly.

Reply:
If the user creates an instance with redisCluster.from_url(), then the url is being parsed and processed to a startup node in the RedisCluster __init__ method, which then passes the startup node to the nodeManager.
So, no matter what - the nodes manager excepts to get startup_nodes. from_url bool is only there to indicate if we need to create all the other cluster's Redis nodes with a costumed connection pool. See NodesManager::create_redis_node